src/main/java/com/amazon/opendistroforelasticsearch/jobscheduler/utils/LockService.java

```diff
@@ -0,0 +1,228 @@
+import org.elasticsearch.action.admin.indices.create.CreateIndexRequest;
+import org.elasticsearch.action.admin.indices.exists.indices.IndicesExistsRequest;
+import org.elasticsearch.action.delete.DeleteRequest;
+import org.elasticsearch.action.delete.DeleteResponse;
+import org.elasticsearch.action.get.GetRequest;
+import org.elasticsearch.action.get.GetResponse;
+import org.elasticsearch.action.index.IndexRequest;
+import org.elasticsearch.action.index.IndexResponse;
+import org.elasticsearch.action.update.UpdateRequest;
+import org.elasticsearch.action.update.UpdateResponse;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.client.IndicesAdminClient;
+import org.elasticsearch.common.xcontent.LoggingDeprecationHandler;
+import org.elasticsearch.common.xcontent.NamedXContentRegistry;
+import org.elasticsearch.common.xcontent.ToXContent;
+import org.elasticsearch.common.xcontent.XContentFactory;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.common.xcontent.XContentType;
+import org.elasticsearch.index.engine.DocumentMissingException;
+import org.elasticsearch.index.engine.VersionConflictEngineException;
+import org.elasticsearch.index.seqno.SequenceNumbers;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.nio.charset.StandardCharsets;
+import java.time.Instant;
+
+public class LockService {
+    private static final Logger logger = LogManager.getLogger(LockService.class);
+    private static final String MAPPING_TYPE = "_doc";
+    private static final String LOCK_INDEX_NAME = ".opendistro-job-scheduler-lock";
+
+    private final IndicesAdminClient indicesAdminClient;
+    private final Client client;
+    private boolean isIndexInitialized;
+
+    // This is used in tests to control time.
+    private Instant testInstant;
+
+    public LockService(final IndicesAdminClient indicesAdminClient, final Client client) {
+        this.indicesAdminClient = indicesAdminClient;
+        this.client = client;
+        this.isIndexInitialized = false;
+    }
+
+    private String lockMapping() {
+        try {
+            InputStream in = LockService.class.getResourceAsStream("opendistro_job_scheduler_lock.json");
+            StringBuilder stringBuilder = new StringBuilder();
+            BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(in, StandardCharsets.UTF_8));
+            for (String line; (line = bufferedReader.readLine()) != null; ) {
+                stringBuilder.append(line);
+            }
+            return stringBuilder.toString();
+        } catch (IOException e) {
+            throw new IllegalArgumentException("Lock Mapping cannot be read correctly.");
+        }
+    }
+
+    @VisibleForTesting
+    boolean createLockIndex() {
+        final boolean exists = indicesAdminClient.exists(new IndicesExistsRequest(LOCK_INDEX_NAME).local(true)).actionGet().isExists();
```

Comment:
Can we pass in the clusterService and just read from cluster state routing table?